db/src/debug.rs

```diff
@@ -12,55 +12,203 @@
-pub fn datoms(conn: &rusqlite::Connection, db: &DB) -> Result<Vec<Datom>> {
-    // TODO: fewer magic numbers!
-    datoms_after(conn, db, &0x10000000)
+impl Datoms {
+    pub fn into_edn_raw<T, U>(&self, tx_id: &T, tx_instant: &U) -> edn::Value
+        where T: Fn(i64) -> edn::Value, U: Fn(i64) -> edn::Value {
+        edn::Value::Set((&self.0).into_iter().map(|x| x.into_edn(tx_id, tx_instant)).collect())
+    }
+
+    pub fn into_edn(&self) -> edn::Value {
+        self.into_edn_raw(&label_tx_id, &label_tx_instant)
+    }
+}
+
+impl Transactions {
+    pub fn into_edn_raw<T, U>(&self, tx_id: &T, tx_instant: &U) -> edn::Value
+        where T: Fn(i64) -> edn::Value, U: Fn(i64) -> edn::Value {
+        edn::Value::Vector((&self.0).into_iter().map(|x| x.into_edn_raw(tx_id, tx_instant)).collect())
+    }
+
+    pub fn into_edn(&self) -> edn::Value {
+        self.into_edn_raw(&label_tx_id, &label_tx_instant)
+    }
+}
+
+/// Convert a numeric entid to an ident `Entid` if possible, otherwise a numeric `Entid`.
+fn to_entid(db: &DB, entid: i64) -> Entid {
+    db.schema.get_ident(&entid).and_then(|ident| to_namespaced_keyword(&ident)).map_or(Entid::Entid(entid), Entid::Ident)
 }
 
-/// Return the set of datoms in the store with transaction ID strictly
-/// greater than the given `tx`, ordered by (tx, e, a, v).
-pub fn datoms_after(conn: &rusqlite::Connection, db: &DB, tx: &i32) -> Result<Vec<Datom>> {
-    let mut stmt: rusqlite::Statement = conn.prepare("SELECT e, a, v, value_type_tag FROM datoms WHERE tx > ? ORDER BY tx, e, a, v")?;
+/// Return the set of datoms in the store, ordered by (e, a, v, tx), but not including any datoms of
+/// the form [... :db/txInstant ...].
+pub fn datoms(conn: &rusqlite::Connection, db: &DB) -> Result<Datoms> {
+    datoms_after(conn, db, bootstrap::TX0 - 1)
+}
+
+/// Return the set of datoms in the store with transaction ID strictly greater than the given `tx`,
+/// ordered by (e, a, v, tx).
+///
+/// The datom set returned does not include any datoms of the form [... :db/txInstant ...].
+pub fn datoms_after(conn: &rusqlite::Connection, db: &DB, tx: i64) -> Result<Datoms> {
+    let mut stmt: rusqlite::Statement = conn.prepare("SELECT e, a, v, value_type_tag, tx FROM datoms WHERE tx > ? ORDER BY e ASC, a ASC, v ASC, tx ASC")?;
+
+    let r: Result<Vec<_>> = stmt.query_and_then(&[&tx], |row| {
+        let e: i64 = row.get_checked(0)?;
+        let a: i64 = row.get_checked(1)?;
+
+        if a == entids::DB_TX_INSTANT {
+            return Ok(None);
+        }
+
+        let v: rusqlite::types::Value = row.get_checked(2)?;
+        let value_type_tag: i32 = row.get_checked(3)?;
+
+        let typed_value = TypedValue::from_sql_value_pair(v, value_type_tag)?;
+        let (value, _) = typed_value.to_edn_value_pair();
+
+        let tx: i64 = row.get_checked(4)?;
+
+        Ok(Some(Datom {
+            e: to_entid(db, e),
+            a: to_entid(db, a),
+            v: value,
```

Comment:
This is true, but the real goal of these debugging routines is to compare to EDN, c.f. #188.  With that in mind, I'm not going to expose `TypedValue` just yet.  (Now, there's `TypedValue::to_edn_value_pair` already, which could be made `Into<edn::Value>` pretty easily, so this might not be much work, but let's see if there's frustration mapping `a` to `Attribute` before we do that work.)